File: rust/src/leetcode/Q22_generate_parentheses.rs
/*
 * @lc app=leetcode id=22 lang=rust
 *
 * [22] Generate Parentheses
 */

// @lc code=start
impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result: Vec<String> = Vec::new();
        // (string, left_count, right_count)
        let mut stack: Vec<(String, i32, i32)> = vec![(String::new(), 0, 0)];

        while let Some((current, left, right)) = stack.pop() {
            if left == n && right == n {
                result.push(current);
                continue;
            }

            if left < n {
                stack.push((current.clone() + "(", left + 1, right));
            }

            if left > right {
                stack.push((current + ")", left, right + 1));
            }
        }

        result
    }
}
// @lc code=end

File: python/Q22_generate_parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#


# @lc code=start
class Solution:
    def backtrack(
        self,
        n: int,
        result: list[str],
        stack: list[str],
        open_count: int = 0,
        close_count: int = 0,
    ) -> None:
        if n == open_count == close_count:
            result.append("".join(stack))
            return

        if open_count < n:
            stack.append("(")
            self.backtrack(n, result, stack, open_count + 1, close_count)
            stack.pop()

        if open_count > close_count:
            stack.append(")")
            self.backtrack(n, result, stack, open_count, close_count + 1)
            stack.pop()

    def generateParenthesis(self, n: int) -> list[str]:
        result: list[str] = []
        stack: list[str] = []

        self.backtrack(n, result, stack)
        return result


# @lc code=end
